src/util/ConfigManager/Validator.h

```diff
@@ -0,0 +1,216 @@
+// Copyright 2023, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Andre Schlegel (September of 2023,
+// schlegea@informatik.uni-freiburg.de)
+
+#pragma once
+
+#include <concepts>
+#include <functional>
+#include <optional>
+#include <stdexcept>
+#include <string>
+
+#include "util/ConfigManager/ConfigOptionProxy.h"
+#include "util/TypeTraits.h"
+
+namespace ad_utility {
+/*
+A validator function is any invocable object, who takes the given parameter
+types, in the same order and transformed to `const type&`, and returns a
+bool.
+*/
+template <typename Func, typename... ParameterTypes>
+concept ValidatorFunction =
+    std::regular_invocable<Func, const ParameterTypes&...> &&
+    isSimilar<std::invoke_result_t<Func, const ParameterTypes&...>, bool>;
+
+// Simple struct, that holds an error message. For use as the return type of
+// invocable object, that fullfill `ExceptionValidator`.
+struct ErrorMessage {
+ private:
+  std::string message_;
+
+ public:
+  // Constructor.
+  explicit ErrorMessage(std::string message) : message_{std::move(message)} {}
+
+  // Getter for the message.
+  const std::string& getMessage() const;
+};
+
+/*
+An exception validator function is any invocable object, who takes the given
+parameter types, in the same order and transformed to `const type&`, and returns
+an instance of `std::optional<ErrorMessage>`.
+*/
+template <typename Func, typename... ParameterTypes>
+concept ExceptionValidatorFunction =
+    std::regular_invocable<Func, const ParameterTypes&...> &&
+    isSimilar<std::invoke_result_t<Func, const ParameterTypes&...>,
```

Comment:
I thiunk you can extract to useful concepts from this that are more generic and can live in `util/TypeTraits.h`:
```
template <typename Fn, typename Ret, typename...Args>
concept InvocableWithSimilarReturnType;
```
And analogously `InvocableWithExactReturnType`

(basically the same as `invocable_r` but, disallowing conversions.

Reply:
I agree with the idea, but shouldn't that be a separate branch, where we also try to integrate it everywhere?

Reply:
Done, see [b2867c7](https://github.com/ad-freiburg/qlever/pull/1118/commits/b2867c75b01c11ff1e7939c33d19c0a6d9c2a854).